core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/waiters/AsyncWaiterHandler.java

```diff
@@ -0,0 +1,118 @@
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.core.internal.waiters;
+
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Supplier;
+import software.amazon.awssdk.annotations.NotThreadSafe;
+import software.amazon.awssdk.annotations.SdkInternalApi;
+import software.amazon.awssdk.core.exception.SdkClientException;
+import software.amazon.awssdk.core.waiters.PollingStrategy;
+import software.amazon.awssdk.core.waiters.WaiterAcceptor;
+import software.amazon.awssdk.core.waiters.WaiterResponse;
+import software.amazon.awssdk.core.waiters.WaiterState;
+import software.amazon.awssdk.utils.Logger;
+import software.amazon.awssdk.utils.Validate;
+
+/**
+ * Handler for async waiter operations
+ * @param <T> the type of the response
+ */
+@SdkInternalApi
+@NotThreadSafe
+public final class AsyncWaiterHandler<T> {
+    private static final Logger log = Logger.loggerFor(AsyncWaiterHandler.class);
+    private final PollingStrategy pollingStrategy;
+    private final ScheduledExecutorService executorService;
+    private final WaiterHandlerHelper<T> handlerHelper;
+    private int attemptNumber = 0;
+
+    public AsyncWaiterHandler(PollingStrategy pollingStrategy,
+                              List<WaiterAcceptor<T>> waiterAcceptors,
+                              ScheduledExecutorService executorService) {
+        this.pollingStrategy = Validate.paramNotNull(pollingStrategy, "pollingStrategy");
+        Validate.paramNotNull(waiterAcceptors, "waiterAcceptors");
+        this.executorService = Validate.paramNotNull(executorService, "executorService");
+        this.handlerHelper = new WaiterHandlerHelper<>(waiterAcceptors, pollingStrategy);
+    }
+
+    /**
+     * Execute the provided async polling function
+     */
+    CompletableFuture<WaiterResponse<T>> execute(Supplier<CompletableFuture<T>> asyncPollingFunction) {
+        log.info(() -> "starting to execute");
+        CompletableFuture<WaiterResponse<T>> future = new CompletableFuture<>();
+        doExecute(asyncPollingFunction, future);
+        return future;
+    }
+
+    private void doExecute(Supplier<CompletableFuture<T>> asyncPollingFunction, CompletableFuture<WaiterResponse<T>> future) {
+        ++attemptNumber;
+        log.info(() -> "runOnce" + attemptNumber);
+        runAsyncPollingFunction(asyncPollingFunction, future);
+    }
+
+    private void runAsyncPollingFunction(Supplier<CompletableFuture<T>> asyncPollingFunction,
+                                         CompletableFuture<WaiterResponse<T>> future) {
+        asyncPollingFunction.get().whenComplete((response, exception) -> {
+            Optional<WaiterState> waiterState = handlerHelper.nextWaiterStateIfMatched(response, exception);
```

Comment:
Any reason we don't have the "max attempts" be an acceptor, instead of special-casing it within this handler?  I realize that means we have to pass something through to allow the acceptor to track "execution context" for a waiter invocation (so that it can keep count), but that seems like a useful feature that better separates the responsibilities of this 'handler'.

Reply:
To me, a `WatierAcceptor` is really an inspector that checks the response or exception returned from an operation and it shouldn't really care about the execution context. IMO, It's the handler's responsibility to handle it, rather than the WaiterAcceptor.

It'd probably be something like the following. Not sure if I like the fact that `WaiterExecutionContext`  is added to be only used by this acceptor and it feels the acceptor is more like a WaiterIntercetor now. 

```
    static <T> WaiterAcceptor<T> retryAttempsAcceptor(int maxAttempts) {
        return new WaiterAcceptor<T>() {
            @Override
            public WaiterState waiterState() {
                return WaiterState.FAILURE;
            }

            @Override
            public boolean matches(T response, WaiterExecutionContext context) {
                if (context.attempts() > maxAttemtps) {
                    return true;
                }
                return false;
            }

            @Override
            public boolean matches(Throwable throwable, WaiterExecutionContext context) {
                if (context.attempts() > maxAttemtps) {
                    return true;
                }
                return false;
            }
        }
    }
```

Reply:
I think I might be trying too hard. You're probably right that we should keep it simple, since I don't see us needing really fancy acceptors. 

My original comment was thinking a little more generic, like:

```java
    static <T> WaiterAcceptor<T> retryAttempsAcceptor(int maxAttempts) {
        return new WaiterAcceptor<T>() {
            private static AttributeKey<Integer> ATTEMPTS = AttributeKey.create("Attempts");

            @Override
            public WaiterState waiterState() {
                return WaiterState.FAILURE;
            }

            @Override
            public boolean matches(T response, WaiterExecutionContext context) {
                incrementAttempt(context);
                if (getAttempts(context) > maxAttempts) {
                    return true;
                }
                return false;
            }

            @Override
            public boolean matches(Throwable throwable, WaiterExecutionContext context) {
                incrementAttempt(context);
                if (getAttempts(context) > maxAttempts) {
                    return true;
                }
                return false;
            }
            
            private void incrementAttempt(WaiterExecutionContext context) {
                context.setAttribute(ATTEMPTS, getAttempts(context) + 1);
            }

            private int getAttempts(WaiterExecutionContext context) {
                Integer attempts = context.getAttribute(ATTEMPTS);
                return attempts == null ? 1 : attempts;
            }
        }
    }
```